### DISCUSSION TITLE
Sending multiple queries/mutations combined in one request.

### DISCUSSION THREAD

**Discussion by renepardon** [spawnia/sailor#90] [Ideas]
If you implement complex views based on microservice (GraphQL Federation), you might end up doing this with spawnia/sailor:

```php
$user = $userService->fetchById($id)->data->getUserById;
$roles = $roleService->fetchAll(null, OrderByClause::make('name', 'asc'));
$timezones = $translationService->fetchTimezones();
$locales = $translationService->fetchLocales();
```

Therefore it would be pretty nice to be able to combine all queries in one request against the API Gateway rather than doing 4 requests.
I could think about using the global Operation class to do something like:

```php
$user = GetUserById::prepareQuery($id)
$rolesQuery = GetUserRoles::prepareQuery(null, OrderByClause::make('name', 'asc'))

$result = Operation::execute($userQuery, $rolesQuery);
```

`$result` would then contain a combined result for the requested user and all available roles. Something similar I could imagine for mutations.

**Comment by spawnia** [comment]
The proposed API looks decent and I see the value in ad-hoc composition of queries in code. However, I presume it might take a lot of tricky implementation and design work - how would this interact with the HTTP client (which is pluggable), with mocking/testing, with the generated code?

Something that works now is to formulate a single large query/mutation that does all the things you need in one go:

```graphql
query UserAndRoles($id: ID!) {
  user(id: $ID) { ... }
  roles { ... }
}
```
